UnitTestingError

when I add the code blow to the LoginPresenterTest.java
```javascript
@Test
public void testInvalidEmailLoginShouldShowError() {
        //invalid email address
        String email = "dummygmail.com";
        String password = "password";
        mLoginPresenter.onServerLoginClick(email, password);
        verify(mMockLoginMvpView).onError(R.string.invalid_email);
}
```
and run the test class LoginPresenterTest, then error occurred:
```
java.lang.ExceptionInInitializerError
	at io.reactivex.android.schedulers.AndroidSchedulers$1.call(AndroidSchedulers.java:35)
	at io.reactivex.android.schedulers.AndroidSchedulers$1.call(AndroidSchedulers.java:33)
	at io.reactivex.android.plugins.RxAndroidPlugins.callRequireNonNull(RxAndroidPlugins.java:70)
	at io.reactivex.android.plugins.RxAndroidPlugins.initMainThreadScheduler(RxAndroidPlugins.java:40)
	at io.reactivex.android.schedulers.AndroidSchedulers.<clinit>(AndroidSchedulers.java:32)
	at com.mindorks.framework.mvp.ui.login.LoginPresenter.onServerLoginClick(LoginPresenter.java:67)
	at com.mindorks.framework.mvp.ui.login.LoginPresenterTest.testServerLoginSuccess(LoginPresenterTest.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.DefaultInternalRunner$1.run(DefaultInternalRunner.java:68)
	at org.mockito.internal.runners.DefaultInternalRunner.run(DefaultInternalRunner.java:74)
	at org.mockito.internal.runners.StrictRunner.run(StrictRunner.java:39)
	at org.mockito.junit.MockitoJUnitRunner.run(MockitoJUnitRunner.java:161)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
Caused by: java.lang.RuntimeException: Method getMainLooper in android.os.Looper not mocked. See http://g.co/androidstudio/not-mocked for details.
	at android.os.Looper.getMainLooper(Looper.java)
	at io.reactivex.android.schedulers.AndroidSchedulers$MainHolder.<clinit>(AndroidSchedulers.java:29)
	... 41 more
```

--- a/app/src/main/java/com/mindorks/framework/mvp/di/module/ActivityModule.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/di/module/ActivityModule.java
@@ -32,6 +32,8 @@ import com.mindorks.framework.mvp.ui.main.MainPresenter;
 import com.mindorks.framework.mvp.ui.splash.SplashMvpPresenter;
 import com.mindorks.framework.mvp.ui.splash.SplashMvpView;
 import com.mindorks.framework.mvp.ui.splash.SplashPresenter;
+import com.mindorks.framework.mvp.utils.rx.AppSchedulerProvider;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import dagger.Module;
 import dagger.Provides;
@@ -64,6 +66,11 @@ public class ActivityModule {
     @Provides
     CompositeDisposable provideCompositeDisposable() {
         return new CompositeDisposable();
+    }
+
+    @Provides
+    SchedulerProvider provideSchedulerProvider(){
+        return new AppSchedulerProvider();
     }
 
     @Provides

--- a/app/src/main/java/com/mindorks/framework/mvp/ui/about/AboutPresenter.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/ui/about/AboutPresenter.java
@@ -17,6 +17,7 @@ package com.mindorks.framework.mvp.ui.about;
 
 import com.mindorks.framework.mvp.data.DataManager;
 import com.mindorks.framework.mvp.ui.base.BasePresenter;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import javax.inject.Inject;
 
@@ -30,7 +31,9 @@ public class AboutPresenter<V extends AboutMvpView> extends BasePresenter<V>
         implements AboutMvpPresenter<V> {
 
     @Inject
-    public AboutPresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
-        super(dataManager, compositeDisposable);
+    public AboutPresenter(DataManager dataManager,
+                          SchedulerProvider schedulerProvider,
+                          CompositeDisposable compositeDisposable) {
+        super(dataManager, schedulerProvider, compositeDisposable);
     }
 }

--- a/app/src/main/java/com/mindorks/framework/mvp/ui/base/BasePresenter.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/ui/base/BasePresenter.java
@@ -30,6 +30,7 @@ import com.mindorks.framework.mvp.R;
 import com.mindorks.framework.mvp.data.DataManager;
 import com.mindorks.framework.mvp.data.network.model.ApiError;
 import com.mindorks.framework.mvp.utils.AppConstants;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import javax.inject.Inject;
 import javax.net.ssl.HttpsURLConnection;
@@ -46,14 +47,17 @@ public class BasePresenter<V extends MvpView> implements MvpPresenter<V> {
     private static final String TAG = "BasePresenter";
 
     private final DataManager mDataManager;
-
+    private final SchedulerProvider mSchedulerProvider;
     private final CompositeDisposable mCompositeDisposable;
 
     private V mMvpView;
 
     @Inject
-    public BasePresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
+    public BasePresenter(DataManager dataManager,
+                         SchedulerProvider schedulerProvider,
+                         CompositeDisposable compositeDisposable) {
         this.mDataManager = dataManager;
+        this.mSchedulerProvider = schedulerProvider;
         this.mCompositeDisposable = compositeDisposable;
     }
 
@@ -82,6 +86,10 @@ public class BasePresenter<V extends MvpView> implements MvpPresenter<V> {
 
     public DataManager getDataManager() {
         return mDataManager;
+    }
+
+    public SchedulerProvider getSchedulerProvider() {
+        return mSchedulerProvider;
     }
 
     public CompositeDisposable getCompositeDisposable() {

--- a/app/src/main/java/com/mindorks/framework/mvp/ui/login/LoginPresenter.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/ui/login/LoginPresenter.java
@@ -22,13 +22,12 @@ import com.mindorks.framework.mvp.data.network.model.LoginRequest;
 import com.mindorks.framework.mvp.data.network.model.LoginResponse;
 import com.mindorks.framework.mvp.ui.base.BasePresenter;
 import com.mindorks.framework.mvp.utils.CommonUtils;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import javax.inject.Inject;
 
-import io.reactivex.android.schedulers.AndroidSchedulers;
 import io.reactivex.disposables.CompositeDisposable;
 import io.reactivex.functions.Consumer;
-import io.reactivex.schedulers.Schedulers;
 
 /**
  * Created by janisharali on 27/01/17.
@@ -40,8 +39,10 @@ public class LoginPresenter<V extends LoginMvpView> extends BasePresenter<V>
     private static final String TAG = "LoginPresenter";
 
     @Inject
-    public LoginPresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
-        super(dataManager, compositeDisposable);
+    public LoginPresenter(DataManager dataManager,
+                          SchedulerProvider schedulerProvider,
+                          CompositeDisposable compositeDisposable) {
+        super(dataManager, schedulerProvider, compositeDisposable);
     }
 
     @Override
@@ -63,8 +64,8 @@ public class LoginPresenter<V extends LoginMvpView> extends BasePresenter<V>
 
         getCompositeDisposable().add(getDataManager()
                 .doServerLoginApiCall(new LoginRequest.ServerLoginRequest(email, password))
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<LoginResponse>() {
                     @Override
                     public void accept(LoginResponse response) throws Exception {
@@ -110,8 +111,8 @@ public class LoginPresenter<V extends LoginMvpView> extends BasePresenter<V>
 
         getCompositeDisposable().add(getDataManager()
                 .doGoogleLoginApiCall(new LoginRequest.GoogleLoginRequest("test1", "test1"))
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<LoginResponse>() {
                     @Override
                     public void accept(LoginResponse response) throws Exception {
@@ -156,8 +157,8 @@ public class LoginPresenter<V extends LoginMvpView> extends BasePresenter<V>
 
         getCompositeDisposable().add(getDataManager()
                 .doFacebookLoginApiCall(new LoginRequest.FacebookLoginRequest("test3", "test4"))
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<LoginResponse>() {
                     @Override
                     public void accept(LoginResponse response) throws Exception {

--- a/app/src/main/java/com/mindorks/framework/mvp/ui/main/MainPresenter.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/ui/main/MainPresenter.java
@@ -20,15 +20,14 @@ import com.mindorks.framework.mvp.data.DataManager;
 import com.mindorks.framework.mvp.data.db.model.Question;
 import com.mindorks.framework.mvp.data.network.model.LogoutResponse;
 import com.mindorks.framework.mvp.ui.base.BasePresenter;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import java.util.List;
 
 import javax.inject.Inject;
 
-import io.reactivex.android.schedulers.AndroidSchedulers;
 import io.reactivex.disposables.CompositeDisposable;
 import io.reactivex.functions.Consumer;
-import io.reactivex.schedulers.Schedulers;
 
 
 /**
@@ -41,8 +40,10 @@ public class MainPresenter<V extends MainMvpView> extends BasePresenter<V>
     private static final String TAG = MainPresenter.class.getSimpleName();
 
     @Inject
-    public MainPresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
-        super(dataManager, compositeDisposable);
+    public MainPresenter(DataManager dataManager,
+                         SchedulerProvider schedulerProvider,
+                         CompositeDisposable compositeDisposable) {
+        super(dataManager, schedulerProvider, compositeDisposable);
     }
 
     @Override
@@ -55,8 +56,8 @@ public class MainPresenter<V extends MainMvpView> extends BasePresenter<V>
         getMvpView().showLoading();
 
         getCompositeDisposable().add(getDataManager().doLogoutApiCall()
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<LogoutResponse>() {
                     @Override
                     public void accept(LogoutResponse response) throws Exception {
@@ -91,8 +92,8 @@ public class MainPresenter<V extends MainMvpView> extends BasePresenter<V>
     public void onViewInitialized() {
         getCompositeDisposable().add(getDataManager()
                 .getAllQuestions()
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<List<Question>>() {
                     @Override
                     public void accept(List<Question> questionList) throws Exception {
@@ -111,8 +112,8 @@ public class MainPresenter<V extends MainMvpView> extends BasePresenter<V>
     public void onCardExhausted() {
         getCompositeDisposable().add(getDataManager()
                 .getAllQuestions()
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .subscribe(new Consumer<List<Question>>() {
                     @Override
                     public void accept(List<Question> questionList) throws Exception {

--- a/app/src/main/java/com/mindorks/framework/mvp/ui/splash/SplashPresenter.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/ui/splash/SplashPresenter.java
@@ -18,15 +18,14 @@ package com.mindorks.framework.mvp.ui.splash;
 import com.mindorks.framework.mvp.R;
 import com.mindorks.framework.mvp.data.DataManager;
 import com.mindorks.framework.mvp.ui.base.BasePresenter;
+import com.mindorks.framework.mvp.utils.rx.SchedulerProvider;
 
 import javax.inject.Inject;
 
 import io.reactivex.ObservableSource;
-import io.reactivex.android.schedulers.AndroidSchedulers;
 import io.reactivex.disposables.CompositeDisposable;
 import io.reactivex.functions.Consumer;
 import io.reactivex.functions.Function;
-import io.reactivex.schedulers.Schedulers;
 
 /**
  * Created by janisharali on 27/01/17.
@@ -36,8 +35,10 @@ public class SplashPresenter<V extends SplashMvpView> extends BasePresenter<V>
         implements SplashMvpPresenter<V> {
 
     @Inject
-    public SplashPresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
-        super(dataManager, compositeDisposable);
+    public SplashPresenter(DataManager dataManager,
+                           SchedulerProvider schedulerProvider,
+                           CompositeDisposable compositeDisposable) {
+        super(dataManager, schedulerProvider, compositeDisposable);
     }
 
     @Override
@@ -48,8 +49,8 @@ public class SplashPresenter<V extends SplashMvpView> extends BasePresenter<V>
 
         getCompositeDisposable().add(getDataManager()
                 .seedDatabaseQuestions()
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(getSchedulerProvider().io())
+                .observeOn(getSchedulerProvider().ui())
                 .concatMap(new Function<Boolean, ObservableSource<Boolean>>() {
                     @Override
                     public ObservableSource<Boolean> apply(Boolean aBoolean) throws Exception {

--- a/app/src/main/java/com/mindorks/framework/mvp/utils/rx/AppSchedulerProvider.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/utils/rx/AppSchedulerProvider.java
@@ -1,0 +1,27 @@
+package com.mindorks.framework.mvp.utils.rx;
+
+import io.reactivex.Scheduler;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ * Created by hrskrs on 5/8/2017.
+ */
+
+public class AppSchedulerProvider implements SchedulerProvider {
+
+  @Override
+  public Scheduler ui() {
+    return AndroidSchedulers.mainThread();
+  }
+
+  @Override
+  public Scheduler computation() {
+    return Schedulers.computation();
+  }
+
+  @Override
+  public Scheduler io() {
+    return Schedulers.io();
+  }
+}

--- a/app/src/main/java/com/mindorks/framework/mvp/utils/rx/SchedulerProvider.java
+++ b/app/src/main/java/com/mindorks/framework/mvp/utils/rx/SchedulerProvider.java
@@ -1,0 +1,16 @@
+package com.mindorks.framework.mvp.utils.rx;
+
+import io.reactivex.Scheduler;
+
+/**
+ * Created by hrskrs on 5/8/2017.
+ */
+
+public interface SchedulerProvider {
+
+  Scheduler ui();
+
+  Scheduler computation();
+
+  Scheduler io();
+}

--- a/app/src/test/java/com/mindorks/framework/mvp/ui/login/LoginPresenterTest.java
+++ b/app/src/test/java/com/mindorks/framework/mvp/ui/login/LoginPresenterTest.java
@@ -18,6 +18,7 @@ package com.mindorks.framework.mvp.ui.login;
 import com.mindorks.framework.mvp.data.DataManager;
 import com.mindorks.framework.mvp.data.network.model.LoginRequest;
 import com.mindorks.framework.mvp.data.network.model.LoginResponse;
+import com.mindorks.framework.mvp.utils.rx.TestSchedulerProvider;
 
 import org.junit.After;
 import org.junit.Before;
@@ -25,17 +26,11 @@ import org.junit.BeforeClass;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Mock;
-import org.mockito.MockitoAnnotations;
 import org.mockito.junit.MockitoJUnitRunner;
 
-import java.util.concurrent.Callable;
-
 import io.reactivex.Observable;
-import io.reactivex.Scheduler;
-import io.reactivex.android.plugins.RxAndroidPlugins;
 import io.reactivex.disposables.CompositeDisposable;
-import io.reactivex.functions.Function;
-import io.reactivex.schedulers.Schedulers;
+import io.reactivex.schedulers.TestScheduler;
 
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.verify;
@@ -50,25 +45,23 @@ public class LoginPresenterTest {
     LoginMvpView mMockLoginMvpView;
     @Mock
     DataManager mMockDataManager;
+
     private LoginPresenter<LoginMvpView> mLoginPresenter;
+    private TestScheduler testScheduler;
 
     @BeforeClass
     public static void onlyOnce() throws Exception {
-        RxAndroidPlugins.setInitMainThreadSchedulerHandler(
-                new Function<Callable<Scheduler>, Scheduler>() {
-                    @Override
-                    public Scheduler apply(Callable<Scheduler> schedulerCallable)
-                            throws Exception {
-                        return Schedulers.trampoline();
-                    }
-                });
     }
 
     @Before
     public void setUp() throws Exception {
-        MockitoAnnotations.initMocks(this);
         CompositeDisposable compositeDisposable = new CompositeDisposable();
-        mLoginPresenter = new LoginPresenter<>(mMockDataManager, compositeDisposable);
+        testScheduler = new TestScheduler();
+        TestSchedulerProvider testSchedulerProvider = new TestSchedulerProvider(testScheduler);
+        mLoginPresenter = new LoginPresenter<>(
+            mMockDataManager,
+            testSchedulerProvider,
+            compositeDisposable);
         mLoginPresenter.onAttach(mMockLoginMvpView);
     }
 
@@ -87,6 +80,8 @@ public class LoginPresenterTest {
 
         mLoginPresenter.onServerLoginClick(email, password);
 
+        testScheduler.triggerActions();
+
         verify(mMockLoginMvpView).showLoading();
         verify(mMockLoginMvpView).hideLoading();
         verify(mMockLoginMvpView).openMainActivity();
@@ -96,7 +91,6 @@ public class LoginPresenterTest {
     @After
     public void tearDown() throws Exception {
         mLoginPresenter.onDetach();
-        RxAndroidPlugins.reset();
     }
 
 }

--- a/app/src/test/java/com/mindorks/framework/mvp/utils/rx/TestSchedulerProvider.java
+++ b/app/src/test/java/com/mindorks/framework/mvp/utils/rx/TestSchedulerProvider.java
@@ -1,0 +1,32 @@
+package com.mindorks.framework.mvp.utils.rx;
+
+import io.reactivex.Scheduler;
+import io.reactivex.schedulers.TestScheduler;
+
+/**
+ * Created by hrskrs on 5/8/2017.
+ */
+
+public class TestSchedulerProvider implements SchedulerProvider{
+
+  private final TestScheduler testScheduler;
+
+  public TestSchedulerProvider(TestScheduler testScheduler) {
+    this.testScheduler = testScheduler;
+  }
+
+  @Override
+  public Scheduler ui() {
+    return testScheduler;
+  }
+
+  @Override
+  public Scheduler computation() {
+    return testScheduler;
+  }
+
+  @Override
+  public Scheduler io() {
+    return testScheduler;
+  }
+}

